File: studio/repositories/working_html_repo.py
"""Project working HTML persistence."""

import time

from studio import db
from studio.models import WorkingHtml, to_doc

_MAX_BYTES = 512 * 1024


class WorkingHtmlError(ValueError):
    pass


def _validate(html):
    if len(html.encode("utf-8")) > _MAX_BYTES:
        raise WorkingHtmlError("working html exceeds 512KB")
    low = html.lower()
    if "<html" not in low or "</html>" not in low:
        raise WorkingHtmlError("working html must be a full document")


def get(project_id):
    from studio import draft

    with db.session() as s:
        row = s.query(WorkingHtml).filter_by(projectId=project_id).first()
        doc = to_doc(row)
    if doc and doc.get("html"):
        doc["html"] = draft.sanitize_html(doc["html"])
    return doc


def put(project_id, html, *, template_id=None):
    from studio import draft

    _validate(html)
    html = draft.sanitize_html(html)
    now = time.time()
    with db.session() as s:
        row = s.query(WorkingHtml).filter_by(projectId=project_id).first()
        if row:
            row.html = html
            row.selectedTemplateId = template_id
            row.updatedAt = now
        else:
            row = WorkingHtml(
                projectId=project_id,
                html=html,
                selectedTemplateId=template_id,
                updatedAt=now,
            )
            s.add(row)
        s.commit()
        return to_doc(row)


def require_html(project_id):
    doc = get(project_id)
    if not doc or not doc.get("html"):
        raise WorkingHtmlError("working html not found")
    return doc["html"]

File: src/server/repos/workingHtmlRepo.ts
import { sanitizeHtml } from "../draft";
import { prisma } from "../db";
import { toDoc } from "./doc";

const MAX_BYTES = 512 * 1024;

export class WorkingHtmlError extends Error {}

function validate(html: string): void {
  if (Buffer.byteLength(html, "utf8") > MAX_BYTES) {
    throw new WorkingHtmlError("working html exceeds 512KB");
  }
  const low = html.toLowerCase();
  if (!low.includes("<html") || !low.includes("</html>")) {
    throw new WorkingHtmlError("working html must be a full document");
  }
}

export async function get(projectId: string): Promise<Record<string, unknown> | null> {
  const row = await prisma().workingHtml.findUnique({ where: { projectId } });
  const doc = toDoc(row) as Record<string, unknown> | null;
  if (doc?.html && typeof doc.html === "string") {
    doc.html = sanitizeHtml(doc.html);
  }
  return doc;
}

export async function put(
  projectId: string,
  html: string,
  templateId: string | null = null,
): Promise<Record<string, unknown>> {
  validate(html);
  const clean = sanitizeHtml(html);
  const now = Date.now() / 1000;
  const row = await prisma().workingHtml.upsert({
    where: { projectId },
    create: { projectId, html: clean, selectedTemplateId: templateId, updatedAt: now },
    update: { html: clean, selectedTemplateId: templateId, updatedAt: now },
  });
  return toDoc(row) as Record<string, unknown>;
}

export async function requireHtml(projectId: string): Promise<string> {
  const doc = await get(projectId);
  if (!doc?.html || typeof doc.html !== "string") {
    throw new WorkingHtmlError("working html not found");
  }
  return doc.html;
}
